### DISCUSSION TITLE
Phone Input for the Form Component

### DISCUSSION THREAD

**Discussion by bartosz-szucko** [shadcn-ui/ui#903] [Ideas]
I think that it would be a good idea to create pre-defined input for phone validation in form components, such as:
![image](https://github.com/shadcn/ui/assets/70040567/aec2ac80-e8b8-41a2-9c4a-1dfc7603709c)
![image](https://github.com/shadcn/ui/assets/70040567/2704188a-1e6b-41a6-b8a8-27225b223bfa)
Currently, there are many libraries for such purposes, but none are friendly for integration and customization - you need to do everything separately for it because of style conflicts.

**Comment by wulzymart** [comment]
This will really be very beneficial, shadcdn is currently gaining popularity, i was hoping this already exists

**Comment by PiotrWarzachowski** [comment]
You are right I might start the project. Recently I fell in love with shadcn. also I wanted to contribute a bit by making this open source project, where I used shadcn , you can contribute and customize some objects if you want, here is my repo : https://github.com/PiotrWarzachowski/DocMind Feel free to check it out, its public on vercel. You can leave a star ⭐ if you want 💓. Have a good year

**Comment by mwolf1989** [comment]
Phone Input would be great!

Maybe wrapped around

https://www.npmjs.com/package/react-phone-number-input

**Comment by vawogbemi** [comment]
Fifth! Yeah I actually really need a cool phone input now so I'm glad this is being talked about.

@shadcn

**Comment by omeralpi** [comment]
Maybe this could be of use to you.

[Shadcn Phone Input](https://github.com/omeralpi/shadcn-phone-input) is a phone input component built as part of the Shadcn design system.

I have also submitted a pull request (#2694) that includes a phone input component.

**Reply by vawogbemi** [reply]
Goated 🐐

**Reply by onurhan1337** [reply]
Nice job 💯

**Reply by fahmiirsyadk** [reply]
thank you so much <3

**Reply by sinoergin** [reply]
fuck me!

**Reply by daudln** [reply]
I am secured.👍👍

**Reply by ricardotech** [reply]
Kudos. Nice job brother.

**Reply by ivnxyz** [reply]
It's amazing! Thank you

**Reply by premiare** [reply]
This is great, keen to get this into the library.

**Reply by Tsandesh** [reply]
please make documents of it 
the documents you provided is just installation

**Reply by codesfromshad** [reply]
@Tsandesh Bro is still cooking, let him cook.

**Reply by bqst** [reply]
Nice job, thank you ! 
Btw, `onChange` should not be required, especially for form action and uncontrolled input.

**Reply by Desgue** [reply]
Right what i needed thank you!

**Reply by omeralpi** [reply]
@bqst `onchange` and `value` were specified as optional. Can you please check the repository again?

**Reply by lash0000** [reply]
thank you very much on this happy holidays! 😃

**Comment by codesfromshad** [comment]
Hey, @omeralpi! I was working on the same problem. Thank you for you contribution. The solution I came up with is a bit janky. I do, however, would like to refine it. Do you want to work on this together?

https://github.com/shadcn-ui/ui/assets/92628452/375b9218-7223-4b57-8d6f-fe5025da48d0

**Reply by lucerowb** [reply]
@codesfromshad, could you share how did you make This input box render email input and Phone no. input change based on the value?

**Reply by codesfromshad** [reply]
Made use of [`libphonenumber-js`](https://www.npmjs.com/package/libphonenumber-js), which ships with a helper function named [`isValidPhoneNumber`](https://www.npmjs.com/package/libphonenumber-js#:~:text=%27TOLL_FREE%27%0A%7D-,Validate%20phone%20number,-import%20%7B%0A%20%20isPossiblePhoneNumber) and used [`zod`](https://www.npmjs.com/package/zod) to test if the provided string is a valid email. Tied that up to `onChange`; then it's just a simple problem of managing the state.

**Reply by gavan-x** [reply]
@codesfromshad very impressive! let me know if you're ever looking for work.

**Comment by chrisidakwo** [comment]
@shadcn is slowly turning into a components library

**Reply by codesfromshad** [reply]
Far from it. `shadcn-ui` gives you complete control over the components, unlike other libraries.

**Reply by chrisidakwo** [reply]
I think I know what shadcn does. I'm just saying with introduction of things like this discussion is asking for, it slowly becomes a component library. Component libraries can also be un-opinionated so the talk of "total control" doesn't really matter.

**Comment by afrinxnahar** [comment]
I made this form component with phone input along with country code using these two libraries together:
 - [React country list](https://www.npmjs.com/package/country-list-with-dial-code-and-flag)
 - [React country flag](https://www.npmjs.com/package/react-country-flag)

And here is the form component: https://github.com/Afrin127329/AlterYouth-homepage/blob/main/src/components/HeroForm.tsx


<img src="https://github.com/user-attachments/assets/22c1cb4d-20f8-4be6-94f7-57a338413acd"/>



       <div className="flex gap-1 justify-center items-center space-y-1.5">
              <Select>
                <SelectTrigger id="phoneInput" className='w-[8rem] h-12 mt-1.5 rounded-md'>
                  <SelectValue defaultValue={"+880"} />
                  <span className="flex items-center gap-1">
                    <ReactCountryFlag className="emojiFlag" countryCode="BD" svg />
                    +880
                  </span>
                </SelectTrigger>
                <SelectContent position="popper" className='w-[250px]'>
                  {countryList.splice(0, 100).map((country) => (
                    <SelectItem key={country.dial_code} value={country.dial_code}>
                      <span className="flex items-center gap-3">
                        <ReactCountryFlag
                          className="emojiFlag"
                          countryCode={country.code}
                          svg
                        />
                        {country.name}{country.dial_code}
                      </span>
                    </SelectItem>
                  ))}
                </SelectContent>
              </Select>
              <Input placeholder='Your Number' className='h-12 rounded-md' />
            </div>

**Reply by devzakir** [reply]
looks okay, needs more improvements
